File: backend/src/main/scala/rider/api/UpdateRiderAvailabilityApi.scala
package rider.api

import domain.shared.given

import cats.effect.IO
import domain.rider.UpdateRiderAvailabilityRequest
import domain.shared.{DeliveryAppState, RiderId, UserRole}
import org.http4s.HttpRoutes
import org.http4s.circe.CirceEntityCodec.*
import rider.app.updateRiderAvailability
import shared.api.routing.*
import shared.app.*

val updateRiderAvailabilityApi: FixedMethodApi1[RiderId, DeliveryAppState] =
  jsonPostApi1[RiderId, UpdateRiderAvailabilityRequest, DeliveryAppState](
    List(routeSegment("api"), routeSegment("delivery"), routeSegment("riders")),
    List(routeSegment("availability")),
  )

File: frontend/src/apis/rider/UpdateRiderAvailabilityApi.ts
import type { DeliveryAppState, RiderId, UpdateRiderAvailabilityRequest } from '@/objects/core/SharedObjects'
import { postNormalizedDeliveryState } from '@/system/api/DeliveryStateClient'
import { buildApiPath1, defineJsonPostApi1, routeSegment } from '@/system/api/TypedApiDefinitions'

export const updateRiderAvailabilityApiDefinition = defineJsonPostApi1<RiderId, UpdateRiderAvailabilityRequest, DeliveryAppState>(
    [routeSegment('api'), routeSegment('delivery'), routeSegment('riders')],
    [routeSegment('availability')],
  )

export function updateRiderAvailability(
  riderId: RiderId,
  payload: UpdateRiderAvailabilityRequest,
) {
  return postNormalizedDeliveryState(buildApiPath1(updateRiderAvailabilityApiDefinition, riderId), payload)
}
